### DISCUSSION TITLE
Obtain email credentials through Aptos keyless

### DISCUSSION THREAD

**Discussion by paulman5** [aptos-labs/aptos-developer-discussions#464] [Dev Doc and Tutorial Feedback]
I have been following several tutorials on implementing Aptos keyless for our application, however some things are still not clear to me. We want to create a mapping from wallet address to email address, but from the aptos TS module useWallet() hook it doesn't seem to have included the email address of a logged in user. 

https://learn.aptoslabs.com/en/code-examples/keyless. This tutorial states that Aptos preserves the privacy of users on purpose, the read me states: 

> Aptos Keyless maintains user privacy in two ways:
> 
> A user’s email address is not revealed on-chain to anybody, including other users and validators.
> A user’s blockchain address and associated transaction history is hidden from the identity provider (e.g. Google).

Does this mean it is not possible to get the user information like the user's email address through Aptos keyless? In the aptos connect dashboard i do see my email listed under "Main account", which can be seen in the screenshot below. 
<img width="1627" alt="Screenshot 2024-09-27 at 21 59 01" src="https://github.com/user-attachments/assets/848b46b1-7321-4b8c-8fbe-bac3a44ee2e8">

My question: Is it impossible to get the email information through Aptos Keyless and if so how does Aptos does it for the Aptos connect dashboard page, do you fetch the information through some other method or still through Keyless?

**Comment by 0x-j** [comment]
You are right, you cannot go from keyless address to email

Aptos connect page asks user to login with Google, so email comes from that login process. It's like if I build a normal web2 app, I ask users to login with Google, I can get the logged in user's email

**Reply by paulman5** [reply]
So does that mean I can't achieve the same when building my DApp with Aptos or do I need some kind of workaround?

**Reply by 0x-j** [reply]
I think you can add sign in with google
https://developers.google.com/identity/gsi/web/guides/display-button

**Reply by paulman5** [reply]
Would that require us to have login with google listed twice, so you log in with keyless and then login again through google? Also thought of maybe an easier way by just letting the user sign in and choose a username and create a mapping address => username.

**Comment by alyssaponzo** [comment]
hey @paulman5 - can you please join our keyless developers TG so that we can help you more easily? https://t.me/+h5CN-W35yUFiYzkx

It is not totally clear to me what you are trying to do as you mention Keyless and refer to the tutorial that shows you how to use the Keyless TS SDK, but your screenshot is of Aptos Connect and mention the wallet adapter, which is a totally different implementation. 

If you use the Keyless TS SDK, it is extremely easy to get the users email as you receive the JWT containing the user's email in plaintext directly from the user's sign in. There are no double sign ins required in either implementation.
